api/sign.js

```diff
@@ -0,0 +1,65 @@
+'use strict';
+var sjcl = require('ripple-lib').sjcl;
+var Seed = require('ripple-lib').Seed;
+var SerializedObject = require('ripple-lib').SerializedObject;
+var validate = require('./lib/validate');
+
+/**
+ * These prefixes are inserted before the source material used to
+ * generate various hashes. This is done to put each hash in its own
+ * "space." This way, two different types of objects with the
+ * same binary data will produce different hashes.
+ *
+ * Each prefix is a 4-byte value with the last byte set to zero
+ * and the first three bytes formed from the ASCII equivalent of
+ * some arbitrary string. For example "TXN".
+ */
+var HASH_TX_ID = 0x54584E00; // 'TXN'
+var HASH_TX_SIGN = 0x53545800; // 'STX'
+var HASH_TX_SIGN_TESTNET = 0x73747800; // 'stx'
+
+function getPublicKey(address, secret) {
+  return Seed.from_json(secret).get_key(address);
+}
+
+function getPublicKeyHex(address, secret) {
+  return getPublicKey(address, secret).to_hex_pub();
+}
+
+function getPrivateKey(address, secret) {
+  return Seed.from_json(secret).get_key(address);
+}
+
+function serialize(txJSON) {
+  return SerializedObject.from_json(txJSON);
+}
+
+function hash(txJSON, prefix) {
+  return serialize(txJSON).hash(prefix || HASH_TX_ID).to_hex();
+}
+
+function signingHash(txJSON, isTestNet) {
+  return hash(txJSON, isTestNet ? HASH_TX_SIGN_TESTNET : HASH_TX_SIGN);
+}
+
+function computeSignature(txJSON, secret) {
+  var privateKey = getPrivateKey(txJSON.Account, secret);
+  var signature = privateKey.sign(signingHash(txJSON));
+  return sjcl.codec.hex.fromBits(signature).toUpperCase();
+}
+
+function sign(txJSON, secret) {
+  validate.txJSON(txJSON);
+  validate.addressAndSecret({address: txJSON.Account, secret: secret});
+
+  if (txJSON.SigningPubKey === undefined) {
+    txJSON.SigningPubKey = getPublicKeyHex(txJSON.Account, secret);
+  }
+  txJSON.TxnSignature = computeSignature(txJSON, secret);
+  return {
+    tx_blob: serialize(txJSON).to_hex(),
```

Comment:
I mean in the final return object block :)  To get the hash, this serialises the json to bytes again. It's already done it for the tx_blob